File: graphs/minimum_spanning/prims/python/prims.py
import heapq

def prims_algorithm(graph, start_node):

    # the adjacency list is already provided for this in main function.

    mst = []
    visited = set()
    cost = 0

    heap = [(0, start_node, None)]

    while heap:
        # [0, 'A', None]
        weight, current_node, previous_node = heapq.heappop(heap)

        if current_node in visited:
            continue
            
        visited.add(current_node)
        if previous_node is not None:
            mst.append((previous_node, current_node, weight))
        
        cost += weight

        for neighbour, n_wight in graph[current_node]:
            if neighbour not in visited:
                heapq.heappush(heap, (n_wight, neighbour, current_node))

    return mst, cost



if __name__ == "__main__":

    graph = {
        'A': [('B', 2), ('C', 3)],
        'B': [('A', 2), ('C', 1), ('D', 4)],
        'C': [('A', 3), ('B', 1), ('D', 5)],
        'D': [('B', 4), ('C', 5)],
    }

    start_node = 'A'
    mst, cost = prims_algorithm(graph, start_node)

    print("Edges in MST:")
    for edge in mst:
        print(f"{edge[0]} - {edge[1]} (Weight: {edge[2]})")

    print(f"Total cost of MST: {cost}") 

File: graphs/minimum_spanning/prims/go/prims.go
// package main

// import "fmt"

// type Edge struct {
// 	weight int32
// 	node string
// 	prev string
// }

// func main() {
// 	graph := map[string][]Edge{
// 		"A": {{weight: 2, node: "B"}, {weight: 3, node: "C"}},
// 		"B": {{weight: 2, node: "A"}, {weight: 1, node: "C"}, {weight: 4, node: "D"}},
// 		"C": {{weight: 3, node: "A"}, {weight: 1, node: "B"}, {weight: 5, node: "D"}},
// 		"D": {{weight: 4, node: "B"}, {weight: 5, node: "C"}},
// 	}

// 	startNode := "A"
// 	mst, cost := primsAlgorithm(graph, startNode)

// 	fmt.Println("Edges in MST:")
// 	for _, edge := range mst {
// 		fmt.Printf("%s - %s (Weight: %d)\n", edge.prev, edge.node, edge.weight)
// 	}

// 	fmt.Printf("Total cost of MST: %d\n", cost)
// }